src/utils/Subscription.js

```diff
@@ -44,13 +44,37 @@ function createListenerCollection() {
   }
 }
 
+// subscriptions are set up in componentDidMount, so we need to delay the actual subscribing until then.
+// To do this, the subscription object is created but does not subscribe until the parent is ready
+// in cDM, a parent calls "hydrate()" which triggers the resolution of the promise. If
+// a child component tries to subscribe before this point, it will wait for the parent subscription
+// to be ready and then try to subscribe.
 export default class Subscription {
   constructor(store, parentSub, onStateChange) {
     this.store = store
     this.parentSub = parentSub
     this.onStateChange = onStateChange
     this.unsubscribe = null
     this.listeners = nullListeners
+    this.loaded = false
+    const resolve = resolve => {
+      this.resolve = () => {
+        this.loaded = true
+        resolve()
+      }
+    }
+    this.ready = new Promise(resolve)
```

Comment:
Per your question: I'm not sure if we've ever actually defined what our target JS environment is.  Then again, we've always back-compiled to plain ES5.

What exactly is going on here with this `resolve()` business?

Is there another way to implement this without using a Promise?

Reply:
You could try using a really minimal fake polyfill, but without being able to use the microtask loop.

```js
const FakePromise = typeof Promise === 'function' ? Promise : class FakePromise {
    constructor (resolver) {
      this._resolved = false
      this._cbs = []
      resolver(this._resolve)
    }

    then (cb) {
      this._cbs.push(cb)
      this._notify()
    }

    _resolve () {
      this._resolved = true
      this._notify()
    }

    _notify () {
      if (!this._resolved) { return }
      const cbs = this._cbs.splice(0)
      for (const cb of cbs) {
        cb()
      }
    }
  }
}
```